src/Storages/ExternalStream/Kafka/Consumer.h

```diff
@@ -33,14 +30,16 @@ class Consumer : boost::noncopyable
 
     void consumeBatch(Topic & topic, Int32 partition, uint32_t count, int32_t timeout_ms, Callback callback, ErrorCallback error_callback) const;
 
+    void shutdown() { stopped = true; }
+
     std::string name() const { return rd_kafka_name(rk.get()); }
 
 private:
     void backgroundPoll(UInt64 poll_timeout_ms) const;
 
     klog::KafkaPtr rk {nullptr, rd_kafka_destroy};
     ThreadPool poller;
-    std::atomic_flag stopped;
+    bool stopped {false};
```

Comment:
multiple thread unsafe ?  one thread call `shutdown`, another thread call `consumeBatch`

Reply:
It appears there is still race

Reply:
No, an atomic is not needed here, there won't be a race.

So, the `stopped` flag is to for telling the `Consumer` (and the `Producer` in the below case) that it's already stopped. And, when we set `stopped` to `true`, we don't do anything else ( like, we don't detroy anything ), so there is nothing to guard with it, i.e. there is no such logic in our code:
```c++
if (!stopped)
{
    stopped = true;
    /// do something
}
else
{
    /// do something else
}
```
So there is no race.

The purpose is to to let operations on the `Consumer` fail, but it does not have to fail immediately. Take your case as an example, if one thread call `shutdown`, and another call `consumeBatch`, it's fine, it does not care which one wins, even `consumeBatch` wins, nothing bad will happen, and next time when `consumeBatch` is called, it will fail as expected, and then, `KafkaSource` will stopped reading more data. And also, we don't even need to guard `shutdown`, it can be called multiple times, no side effects.

Please let me know if this does not make sense to you.

Reply:
Regardless it does not have any impact in current implementation, I reverted it back, and now use `std::atomic_flag`.

Reply:
The race i meant it not the stopped data member but at high level..
1. Client call `isStopped()`, returns false, context switch to 2)
2. Consumer then stopped, switch back to 1)
3. Client already checked consumer is not stopped, but it actually is, continue operating on stopped consumer ?

Reply:
Same for producer and we run into this problem before